File: liblemonshark/examples/field_description/field_description_demo.c
/*
Copyright (c) 2024 DevAM. All Rights Reserved.

SPDX-License-Identifier: GPL-2.0-only
*/

#include "lemonshark.h"

static void handle_error(const char *error_message)
{
    g_print("%s", error_message);
}

static void print_field_description(field_description_t* field_description)
{
    gint32 field_type = ls_field_description_type_get(field_description);
    gint32 field_id = ls_field_description_id_get(field_description);
    const char* name = ls_field_description_name_get(field_description);
    const char* display_name = ls_field_description_display_name_get(field_description);
    const char* type_name = ls_field_type_get_name(field_type);
    gint32 parent_id = ls_field_description_parent_id_get(field_description);

    g_print("%s (Name: %s, Type: %s, Id: %i", display_name, name, type_name, field_id);
    if (parent_id >= 0)
    {
        g_print(", Parent: %i", parent_id);
    }
    g_print(")\n");
}

int main(void)
{
    void *error_handler = ls_error_handler_add(handle_error);

    // Set the environment variable LS_EXAMPLE_FILE to the path of a valid trace file before running this example.
    const char *trace_file_path = g_getenv("LS_EXAMPLE_FILE");

    char *error_message = NULL;
    gint32 init_result = ls_session_create_from_file(trace_file_path, "", NULL, &error_message);

    if (init_result == LS_ERROR)
    {
        g_print("%s", error_message);
        g_free(error_message);
        return -1;
    }

    // Get all field descriptions
    gint32 field_descriptions_count = 0;
    field_description_t** field_descriptions = ls_field_description_get_all(&field_descriptions_count);

    for (gint32 i = 0; i < field_descriptions_count; i++)
    {
        field_description_t* field_description = field_descriptions[i];

        print_field_description(field_description);

        ls_field_description_free(field_description);
    }

    ls_memory_free(field_descriptions);

    g_print("\n");

    // Get a field description by name
    field_description_t* frame_length_field_description = ls_field_description_get_by_name("frame.len");
    print_field_description(frame_length_field_description);
    ls_field_description_free(frame_length_field_description);

    ls_session_close();

    ls_error_handler_remove(error_handler);

    return 0;
}

File: python/examples/field_description/field_description_demo.py
"""
Copyright (c) 2024 DevAM. All Rights Reserved.

SPDX-License-Identifier: GPL-2.0-only
"""

import sys
import os
import platform
from typing import *

system: str = platform.system().lower()
architecture: str = platform.machine().lower()

lemonshark_directory: str = os.path.dirname(os.path.realpath(__file__)) + f"/../../lemonshark"
lemonshark_directory = os.path.realpath(lemonshark_directory)
sys.path.append(lemonshark_directory)

from lemonshark import LemonShark, Session, FieldDescription, FieldType


def print_field_description(field_description: FieldDescription) -> None:
    display_name: str = field_description.get_display_name()
    name: str = field_description.get_name()
    type: int = field_description.get_type()
    id: int = field_description.get_id()
    parent_id: int = field_description.get_parent_id()
    type_name: str = FieldType.get_name(type)

    print(f"""{display_name} (Name: {name}, Type: {type_name}, Id: {id})""", end="")
    if parent_id >= 0:
        print(f""", Parent: {parent_id}""", end="")
    print(")")


def main():
    wireshark_directory: str = None
    if system == "windows":
        if architecture == "amd64":
            wireshark_directory = os.path.dirname(os.path.realpath(__file__)) + f"/../../../build/wireshark/windows/amd64/run/RelWithDebInfo"
            wireshark_directory = os.path.realpath(wireshark_directory)

    LemonShark.init([wireshark_directory])

    trace_file_path: str = os.environ["LS_EXAMPLE_FILE"]

    session: Session = Session.create_from_file(trace_file_path, "")

    try:
        # Get all field descriptions
        field_descriptions: List[FieldDescription] = FieldDescription.get_all()

        for field_description in field_descriptions:
            print_field_description(field_description)

        print("")
        
        # Get a field description by name        
        frame_field_description: FieldDescription = FieldDescription.get_by_name("frame")
        print_field_description(frame_field_description)

    except Exception as ex:
        print(ex)
    finally:
        session.close()


if __name__ == "__main__":
    main()
